### ISSUE TITLE
Support for CORS

### ISSUE THREAD

**Issue by sp3ctum** [KamWithK/AnkiconnectAndroid#13]
Hi, I'm working on my hobby project which is a Japanese web dictionary called hare: https://github.com/sp3ctum/hare

It's nothing special but I use it in my own workflow. I'd like to add support for anki connect as well as anki connect android in order to quickly add cards when I'm reading.

Anki connect allows customizing the CORS origin list in its config, here is an example:
```json
{
    "apiKey": null,
    "apiLogPath": null,
    "ignoreOriginList": [],
    "webBindAddress": "127.0.0.1",
    "webBindPort": 8765,
    "webCorsOriginList": [
        "http://localhost",
        "https://sakura-paris.org"
    ]
}
```

https://sakura-paris.org is where I run my dictionary. Without the added configuration, all requests are blocked by chrome because they target another host. This is what it looks like:
![image](https://user-images.githubusercontent.com/300791/208239642-cd0a8b08-a9fe-48e7-969d-34512bcad42b.png)

But after this configuration has been saved, I can access the service from chrome. 
![image](https://user-images.githubusercontent.com/300791/208239526-0730be68-127a-47ca-984b-52c72eca3223.png)
![image](https://user-images.githubusercontent.com/300791/208239541-427dfa2b-0b5f-4be1-9079-85ded9678327.png)

I'll start by working on anki connect desktop but I understand AnkiConnectAndroid uses the same api (roughly) so I think I might be able to support both. This, I imagine, is what yomichan does as well.

Do you think it could be possible to add support for configurable CORS origins? I could maybe submit a PR for this but I'm not exactly sure how to do it yet.

**Comment by KamWithK** [comment]
It's probably possible I guess but I'm not sure how to do it either

**Comment by sp3ctum** [comment]
Will you accept a PR for it? I might be able to implement it, but no promise on any deadlines of course.

**Comment by KamWithK** [comment]
Yeah sure

**Comment by Aquafina-water-bottle** [comment]
> Will you accept a PR for it? I might be able to implement it, but no promise on any deadlines of course.

I was actually playing around with this earlier to see if I could get it work with my Anki template within Ankidroid. I believe this is what you're looking for:

```diff
diff --git a/app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java b/app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java
index 908622e..fd3e2c1 100644
--- a/app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java
+++ b/app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java
@@ -52,6 +52,14 @@ public class RouteHandler extends RouterNanoHTTPD.DefaultHandler {
             e.printStackTrace();
         }
 
-        return apiHandler.chooseAPI(files.get("postData"), session.getParameters());
+        NanoHTTPD.Response rep = null;
+        if (files.get("postData") == null) {
+            rep = newFixedLengthResponse("Ankiconnect Android is running.");
+        } else {
+            rep = apiHandler.chooseAPI(files.get("postData"), session.getParameters());
+        }
+
+        rep.addHeader("Access-Control-Allow-Origin", "http://localhost");
+        return rep;
     }
```

This hard codes the localhost header into the program, but I never looked into how to make a GUI for it. I'll let you pick it up from here if you're interested! @sp3ctum

**Comment by Aquafina-water-bottle** [comment]
I should also mention @sp3ctum that the most recent PR I made (#16) implements a basic settings section on the toolbar if you want to work from there.

**Comment by sp3ctum** [comment]
Cool, that looks great @Aquafina-water-bottle , should be a nice place to add some kind of "Allowed hosts/sites" setting section.

I played around with your code snippet and was able to get it to work with the headers below. There are still some issues with some requests that I haven't solved so this shouldn't be considered a final solution by any means.
```java
        rep.addHeader("Access-Control-Allow-Origin", "*"); // will need configuration, probably not good to allow all origins
        rep.addHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
        rep.addHeader("Access-Control-Allow-Methods ", "GET, POST, PUT, OPTIONS");
```

Also I opened some issues for some minor things here in this repo.

I am in the process of changing jobs in the coming weeks and don't have the means to finish this right now.. But I still need it done later for my app, so I think I'll take a look at this once things become somewhat settled down.
